url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/syntax/Syntax.java

```diff
@@ -0,0 +1,262 @@
+			}
+			else	{
+				if (this.rules == null)	{
+					allocateRules(rules);
+				}
+			}
+		}
+	}
+	
+	/** Returns the number of rules of this Syntax. */
+	public int size()	{
+		return rules != null ? rules.size() : 0;
+	}
+	
+	/** Returns the rule at the given position. */
+	public Rule getRule(int i)	{
+		return (Rule) rules.get(i);
+	}
+	
+	/** Appends a new rule to this Syntax. */
+	public void addRule(Rule rule)	{
+		insertRule(size(), rule);
+	}
+	
+	/** Inserts a new rule into this Syntax (needed to set artificial START rule at position 0). */
+	public void insertRule(int i, Rule rule)	{
+		if (rules == null)	{
+			allocateRules(null);
+		}
+		
+		if (ruleHash.get(rule) != null)
+			return;	// already contained
+		
+		ruleHash.put(rule, rule);
+		rules.add(i, rule);
+	}
+	
+	/** Removes the rule at passed index from this Syntax. */
+	public void removeRule(int index)	{
+		rules.remove(index);
+	}
+	
+
+	private void allocateRules(List rules)	{
+		int size = rules == null || rules.size() <= 0 ? 64 : rules.size();
+		this.rules = new ArrayList(size);
+		this.ruleHash = new Hashtable(size);
+	}
+
+
+	/** Returns a List of top level Rules (that do not appear on right side of any contained rule). */
+	public List findStartRules()	{
+		List roots = new ArrayList(1);
+		
+		for (int i = 0; i < size(); i++)	{
+			Rule rule = getRule(i);
+			String nonterm = rule.getNonterminal();
+
+			// search this left side nonterminal on right sides of all rules
+			boolean found = nonterm.equals(Token.TOKEN) || nonterm.equals(Token.IGNORED);
+			for (int j = 0; found == false && j < size(); j++)	{
+				Rule r = getRule(j);
+				if (j != i && r.getNonterminal().equals(nonterm) == false)	// start rule is allowed to be recursive
+					for (int k = 0; found == false && k < r.rightSize(); k++)
+						if (r.getRightSymbol(k).equals(nonterm) && false == r.getNonterminal().equals(Token.TOKEN) && false == r.getNonterminal().equals(Token.IGNORED))
+							found = true;
+			}
+			
+			if (found == false)	// must be a top level rule
+				roots.add(rule);
+		}
+		
+		return roots;
+	}
+
+
+	/**
+		Simplify the syntax. Search rules that have a nonterminal that is defined only once
+		and contains exactly one symbol on right side. Such a rule gets deleted and its references
+		get substituted by the symbol on right side of that rule.
+		<p>
+		This method is called by default from SerializedObject (base class for all builders).
+		<p>
+		CAUTION: Such symbols will not be called within Semantic!
+	*/
+	public void resolveSingulars()	{
+		Rule singular;
+		while ((singular = removeSingular()) != null)	{
+			String substitute = singular.getNonterminal();
+
+			for (int i = 0; i < size(); i++)	{	// substitute rules containing the singular
+				Rule rule = getRule(i);
+
+				for (int j = 0; j < rule.rightSize(); j++)
+					if (rule.getRightSymbol(j).equals(substitute))
+						rule.setRightSymbol(singular.getRightSymbol(0), j);
+			}
+		}
+	}
+
+	private Rule removeSingular()	{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.